File: ios/Runner/BarometerChannel.swift
import Foundation
import CoreMotion
import Flutter

class BarometerChannel: NSObject, FlutterStreamHandler {
    static let CHANNEL = "com.samir.gotrek/barometer"

    private let altimeter = CMAltimeter()
    private var eventSink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events

        guard CMAltimeter.isRelativeAltitudeAvailable() else {
            events(FlutterError(code: "SENSOR_NOT_FOUND", message: "Barometer sensor not available", details: nil))
            return nil
        }

        altimeter.startRelativeAltitudeUpdates(to: .main) { [weak self] data, error in
            if let error = error {
                self?.eventSink?(FlutterError(code: "SENSOR_ERROR", message: error.localizedDescription, details: nil))
                return
            }
            if let data = data {
                // Send pressure in hPa (kPa * 10 = hPa)
                let pressureHPa = data.pressure.doubleValue * 10.0
                self?.eventSink?(pressureHPa)
            }
        }

        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        altimeter.stopRelativeAltitudeUpdates()
        eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/com/samir/gotrek/gotrek/BarometerChannel.kt
package com.samir.gotrek.gotrek

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import io.flutter.plugin.common.EventChannel

class BarometerChannel(private val context: Context) : EventChannel.StreamHandler {

    companion object {
        const val CHANNEL = "com.samir.gotrek/barometer"
    }

    private var sensorManager: SensorManager? = null
    private var pressureSensor: Sensor? = null
    private var eventSink: EventChannel.EventSink? = null

    private val sensorListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            if (event.sensor.type == Sensor.TYPE_PRESSURE) {
                // Send pressure in hPa (hectopascals)
                eventSink?.success(event.values[0].toDouble())
            }
        }

        override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
    }

    override fun onListen(arguments: Any?, sink: EventChannel.EventSink) {
        eventSink = sink
        sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        pressureSensor = sensorManager?.getDefaultSensor(Sensor.TYPE_PRESSURE)

        if (pressureSensor == null) {
            sink.error("SENSOR_NOT_FOUND", "Barometer sensor not available", null)
            return
        }

        sensorManager?.registerListener(
            sensorListener,
            pressureSensor,
            SensorManager.SENSOR_DELAY_UI,
        )
    }

    override fun onCancel(arguments: Any?) {
        sensorManager?.unregisterListener(sensorListener)
        eventSink = null
    }
}
